src/librustc_codegen_ssa/mir/block.rs

```diff
@@ -21,766 +23,870 @@ use super::place::PlaceRef;
+                let index = self.codegen_operand(&mut bx, index).immediate();
+
+                let file_line_col = bx.const_struct(&[filename, line, col], false);
+                let file_line_col = bx.static_addr_of(
+                    file_line_col,
+                    align,
+                    Some("panic_bounds_check_loc")
+                );
+                (lang_items::PanicBoundsCheckFnLangItem,
+                 vec![file_line_col, index, len])
             }
+            _ => {
+                let str = msg.description();
+                let msg_str = Symbol::intern(str).as_str();
+                let msg_str = bx.const_str_slice(msg_str);
+                let msg_file_line_col = bx.const_struct(
+                    &[msg_str, filename, line, col],
+                    false
+                );
+                let msg_file_line_col = bx.static_addr_of(
+                    msg_file_line_col,
+                    align,
+                    Some("panic_loc")
+                );
+                (lang_items::PanicFnLangItem,
+                 vec![msg_file_line_col])
+            }
+        };
 
-            mir::TerminatorKind::Assert { ref cond, expected, ref msg, target, cleanup } => {
-                let cond = self.codegen_operand(&mut bx, cond).immediate();
-                let mut const_cond = bx.const_to_opt_u128(cond, false).map(|c| c == 1);
-
-                // This case can currently arise only from functions marked
-                // with #[rustc_inherit_overflow_checks] and inlined from
-                // another crate (mostly core::num generic/#[inline] fns),
-                // while the current crate doesn't use overflow checks.
-                // NOTE: Unlike binops, negation doesn't have its own
-                // checked operation, just a comparison with the minimum
-                // value, so we have to check for the assert message.
-                if !bx.check_overflow() {
-                    if let mir::interpret::EvalErrorKind::OverflowNeg = *msg {
-                        const_cond = Some(expected);
-                    }
-                }
+        // Obtain the panic entry point.
+        let def_id = common::langcall(bx.tcx(), Some(span), "", lang_item);
+        let instance = ty::Instance::mono(bx.tcx(), def_id);
+        let fn_ty = bx.fn_type_of_instance(&instance);
+        let llfn = bx.get_fn(instance);
 
-                // Don't codegen the panic block if success if known.
-                if const_cond == Some(expected) {
-                    funclet_br(self, &mut bx, target);
-                    return;
-                }
+        // Codegen the actual panic invoke/call.
+        helper.do_call(self, &mut bx, fn_ty, llfn, &args, None, cleanup);
+    }
 
-                // Pass the condition through llvm.expect for branch hinting.
-                let cond = bx.expect(cond, expected);
+    fn codegen_call_terminator<'b>(
```

Comment:
This function is still ~450 LOC. We might want to try to refactor this function at some point.

Reply:
Yeah. Leave that to another PR though.